processor-api/src/main/java/com/synopsys/integration/alert/processor/api/extract/ProviderMessageExtractor.java

```diff
@@ -22,10 +22,52 @@
  */
 package com.synopsys.integration.alert.processor.api.extract;
 
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import com.synopsys.integration.alert.processor.api.detail.ProviderMessageHolder;
 import com.synopsys.integration.alert.processor.api.filter.model.FilterableNotificationWrapper;
+import com.synopsys.integration.blackduck.api.manual.component.NotificationContentComponent;
+import com.synopsys.integration.blackduck.api.manual.enumeration.NotificationType;
+
+public abstract class ProviderMessageExtractor<T extends NotificationContentComponent> {
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final NotificationType notificationType;
+    private final Class<T> notificationContentClass;
+
+    protected ProviderMessageExtractor(NotificationType notificationType, Class<T> notificationContentClass) {
+        this.notificationType = notificationType;
+        this.notificationContentClass = notificationContentClass;
+    }
+
+    public NotificationType getNotificationType() {
+        return notificationType;
+    }
+
+    public final ProviderMessageHolder extract(FilterableNotificationWrapper filteredNotification) {
+        if (!notificationContentClass.isAssignableFrom(filteredNotification.getNotificationContentClass())) {
+            logger.error("The notification type provided is incompatible with this extractor: {}", filteredNotification.extractNotificationType());
+            return ProviderMessageHolder.empty();
+        }
+
+        return castToStrongType(filteredNotification.getNotificationContent())
+                   .map(content -> extract(filteredNotification, content))
+                   .orElse(ProviderMessageHolder.empty());
+    }
+
+    protected abstract ProviderMessageHolder extract(FilterableNotificationWrapper filteredNotification, T notificationContent);
 
-public interface ProviderMessageExtractor {
-    ProviderMessageHolder extract(FilterableNotificationWrapper filteredNotification);
+    private Optional<T> castToStrongType(NotificationContentComponent notificationContent) {
+        try {
+            T stronglyTypedContent = notificationContentClass.cast(notificationContent);
+            return Optional.of(stronglyTypedContent);
+        } catch (ClassCastException e) {
+            logger.error("Failed to cast the notification content class from {} to {}", notificationContent.getClass().getSimpleName(), notificationContentClass.getSimpleName(), e);
+            return Optional.empty();
+        }
```

Comment:
I think this is a case of us wanting to fail fast, if we have code that doesn't match expected inputs we should know about it immediately.

Reply:
Hiding it in a log message may mean it sneaks into production

Reply:
You're right, I think this would be a bug. I'll get rid of the exception handling.